### PR TITLE
Crypto 3.1: Fix #208: Added IV to ECIES

### LINKED ISSUE #208: Crypto 3.1: ECIES with non-zero IV
This enhancement adds support for ECIES mode with non-zero initialization vector. See wultra/powerauth-crypto#307 ticket for more details.

### PR DESCRIPTION
**Pull request by hvge**

### CODE REVIEW

File: powerauth-restful-security-spring/src/main/java/io/getlime/security/powerauth/rest/api/spring/service/v3/ActivationService.java [hunk 1]
```diff
@@ -101,17 +101,23 @@ public ActivationLayer1Response createActivation(ActivationLayer1Request request
             final String ephemeralPublicKey = activationData.getEphemeralPublicKey();
             final String encryptedData = activationData.getEncryptedData();
             final String mac = activationData.getMac();
+            final String nonce = activationData.getNonce();
             final Map<String, Object> customAttributes = request.getCustomAttributes();
             final Map<String, String> identity = request.getIdentityAttributes();
 
+            // Validate inner encryption
+            if (nonce == null && !"3.0".equals(eciesContext.getVersion())) {
+                throw new PowerAuthActivationException();
```
Comment (romanstrobl) [inline powerauth-restful-security-spring/src/main/java/io/getlime/security/powerauth/rest/api/spring/service/v3/ActivationService.java]: Missing logging.

File: powerauth-restful-security-javaee/src/main/java/io/getlime/security/powerauth/rest/api/jaxrs/service/v3/ActivationService.java [hunk 2]
```diff
@@ -87,17 +87,23 @@ public ActivationLayer1Response createActivation(ActivationLayer1Request request
             final String ephemeralPublicKey = activationData.getEphemeralPublicKey();
             final String encryptedData = activationData.getEncryptedData();
             final String mac = activationData.getMac();
+            final String nonce = activationData.getNonce();
             final Map<String, Object> customAttributes = request.getCustomAttributes();
             final Map<String, String> identity = request.getIdentityAttributes();
 
+            // Validate inner encryption
+            if (nonce == null && !"3.0".equals(eciesEncryption.getContext().getVersion())) {
+                throw new PowerAuthActivationException();
```
Comment (romanstrobl) [inline powerauth-restful-security-javaee/src/main/java/io/getlime/security/powerauth/rest/api/jaxrs/service/v3/ActivationService.java]: Missing logging.

File: powerauth-restful-security-base/src/main/java/io/getlime/security/powerauth/rest/api/base/provider/PowerAuthEncryptionProviderBase.java [hunk 3]
```diff
@@ -111,15 +111,20 @@
...
             final String mac = eciesRequest.getMac();
+            final String nonce = eciesRequest.getNonce();
 
-            // Verify ECIES request data
+            // Verify ECIES request data. Nonce is required for protocol 3.1+
             if (ephemeralPublicKey == null || encryptedData == null || mac == null) {
                 throw new PowerAuthEncryptionException("Invalid ECIES request data");
             }
+            if (nonce == null && !"3.0".equals(encryptionContext.getVersion())) {
+                throw new PowerAuthEncryptionException("Missing nonce in ECIES request data");
```
Comment (romanstrobl) [inline powerauth-restful-security-base/src/main/java/io/getlime/security/powerauth/rest/api/base/provider/PowerAuthEncryptionProviderBase.java]: Missing logging.

Comment (hvge) [reply powerauth-restful-security-base/src/main/java/io/getlime/security/powerauth/rest/api/base/provider/PowerAuthEncryptionProviderBase.java]: I want to be consistent with the rest of the code. There's no logging just a few lines above in the same situation.

Comment (romanstrobl) [reply powerauth-restful-security-base/src/main/java/io/getlime/security/powerauth/rest/api/base/provider/PowerAuthEncryptionProviderBase.java]: Ok

File: powerauth-restful-security-base/src/main/java/io/getlime/security/powerauth/rest/api/base/provider/PowerAuthEncryptionProviderBase.java [hunk 4]
```diff
@@ -111,15 +111,20 @@
             final String ephemeralPublicKey = eciesRequest.getEphemeralPublicKey();
             final String encryptedData = eciesRequest.getEncryptedData();
             final String mac = eciesRequest.getMac();
+            final String nonce = eciesRequest.getNonce();
 
-            // Verify ECIES request data
+            // Verify ECIES request data. Nonce is required for protocol 3.1+
```
[the remark this answers is not in the document]
Comment (hvge) [reply powerauth-restful-security-base/src/main/java/io/getlime/security/powerauth/rest/api/base/provider/PowerAuthEncryptionProviderBase.java]: I've added info to the class documentation.

**Pull request merged**